README
I'm empty!


Changes
Revision history for Perl extension DBIx::Class::DateTime::Epoch.

0.09 2012-02-01
 - added a test dependency on DateTime::Format::SQLite (RT #74576)
 - added some documentation about the installation requirement of an 
   appropriate DateTime::Format::* module for your RDBMS of choice

0.08 2011-11-21
 - remove warning in case of datetime column without inflate_datetime
   (Alexandr Ciornii)

0.07 2011-01-19
 - fix inflation/deflation routines so that non-epoch datetime columns are
   handled correctly (woosley)

0.06 2009-01-28
 - make sure regex for "int" is case insensitive (RT #54053)

0.05 2009-06-01
 [THING THAT WILL NO LONGER BREAK YOUR CODE]
 - added back-compat for the "epoch => 1", etc syntax

0.04 2009-05-26
 [ENHANCEMENTS]
 - switch to Module::Install

 [THINGS THAT WILL BREAK YOUR CODE]
 - refactor module to itegrate with InflateColumn::DateTime and TimeStamp.
   Requires DBIx::Class 0.08103 and DBIx::Class::TimeStamp 0.07. Please
   read the new documentation.

0.03 2007-04-25
 - don't override user-supplied values for update()

0.02 2006-07-17
 - Re-release onto CPAN

0.01 2006-03-22
 - original version;


t/lib/MySchema/Foo.pm
package MySchema::Foo;

use strict;
use warnings;

use base qw( DBIx::Class );

__PACKAGE__->load_components( qw( DateTime::Epoch TimeStamp Core ) );
__PACKAGE__->table( 'foo' );
__PACKAGE__->add_columns(
    id => {
        data_type         => 'bigint',
        is_auto_increment => 1,
        is_nullable       => 0,
    },
    name => {
        data_type   => 'varchar',
        size        => 10,
        is_nullable => 1,
    },
    bar => { # epoch stored as an int
        data_type        => 'bigint',
        inflate_datetime => 1,
    },
    baz => { # epoch stored as a string
        data_type        => 'varchar',
        size             => 50,
        inflate_datetime => 'epoch',
    },
    # working in conjunction with DBIx::Class::TimeStamp
    creation_time => {
        data_type        => 'bigint',
        inflate_datetime => 1,
        set_on_create    => 1,
    },
    modification_time => {
        data_type        => 'bigint',
        inflate_datetime => 1,
        set_on_create    => 1,
        set_on_update    => 1,
    },
    update_time => {
        data_type        => 'datetime',
        set_on_create    => 1,
        set_on_update    => 1,
        is_nullable      => 1,
    },

);

__PACKAGE__->set_primary_key( 'id' );

1;


lib/DBIx/Class/DateTime/Epoch.pm
package DBIx::Class::DateTime::Epoch;

use strict;
use warnings;

our $VERSION = '0.09';

use base qw( DBIx::Class );

use DateTime;

__PACKAGE__->load_components( qw( InflateColumn::DateTime ) );

# back compat
sub add_columns {
    my( $class, @cols ) = @_;
    my @columns;

    while (my $col = shift @cols) {
        my $info = ref $cols[0] ? shift @cols : {};

        if( my $type = delete $info->{ epoch } ) {
            $info->{ inflate_datetime } = 'epoch';

            if( $type =~ m{^[cm]time$} ) {
                __PACKAGE__->load_components( 'TimeStamp' );
                $info->{ set_on_create } = 1;
                $info->{ set_on_update } = 1 if $type eq 'mtime';
            }
        }

        push @columns, $col => $info;

    }

    $class->next::method( @columns );
}

sub _inflate_to_datetime {
    my( $self, $value, $info, @rest ) = @_;
    return $self->next::method( $value, $info, @rest )
        unless $info->{ data_type } =~ m{int}i || (exists $info->{ inflate_datetime } && $info->{ inflate_datetime } eq 'epoch');

    return DateTime->from_epoch( epoch => $value );
}

sub _deflate_from_datetime {
    my( $self, $value, $info, @rest ) = @_;
    return $self->next::method( $value, $info, @rest )
        unless $info->{ data_type } =~ m{int}i || (exists $info->{ inflate_datetime } && $info->{ inflate_datetime } eq 'epoch');

    return $value->epoch;
}

1;

__END__

=head1 NAME

DBIx::Class::DateTime::Epoch - Automatic inflation/deflation of epoch-based columns to/from DateTime objects

=head1 SYNOPSIS

    package MySchema::Foo;
    
    use base qw( DBIx::Class );
    
    __PACKAGE__->load_components( qw( DateTime::Epoch TimeStamp Core ) );
    __PACKAGE__->add_columns(
        name => {
            data_type => 'varchar',
            size      => 10,
        },
        bar => { # epoch stored as an int
            data_type        => 'bigint',
            inflate_datetime => 1,
        },
        baz => { # epoch stored as a string
            data_type        => 'varchar',
            size             => 50,
            inflate_datetime => 'epoch',
        },
        # working in conjunction with DBIx::Class::TimeStamp
        creation_time => {
            data_type        => 'bigint',
            inflate_datetime => 1,
            set_on_create    => 1,
        },
        modification_time => {
            data_type        => 'bigint',
            inflate_datetime => 1,
            set_on_create    => 1,
            set_on_update    => 1,
        }
    );

=head1 DATETIME::FORMAT DEPENDENCY

There have been no assumptions made as to what RDBMS you will be using. As per 
the note in the L<DBIx::Class::InflateColumn::DateTime documentation|DBIx::Class::InflateColumn::DateTime/DESCRIPTION>, 
you will need to install the DateTime::Format::* module that matches your RDBMS 
of choice.

=head1 DESCRIPTION

This module automatically inflates/deflates DateTime objects from/to epoch
values for the specified columns. This module is essentially an extension to
L<DBIx::Class::InflateColumn::DateTime> so all of the settings, including
C<locale> and C<timezone>, are also valid.

A column will be recognized as an epoch time given one of the following scenarios:

=over 4

=item * C<data_type> is an C<int> of some sort and C<inflate_datetime> is also set to a true value

=item * C<data_type> is some other value (e.g. C<varchar>) and C<inflate_datetime> is explicitly set to C<epoch>.

=back

L<DBIx::Class::TimeStamp> can also be used in conjunction with this module to support
epoch-based columns that are automatically set on creation of a row and updated subsequent
modifications.

=head1 METHODS

=head2 add_columns( )

Provides backwards compatibility with the older DateTime::Epoch API.

=head2 _inflate_to_datetime( )

Overrides column inflation to use C<Datetime-E<gt>from_epoch>.

=head2 _deflate_from_datetime( )

Overrides column deflation to call C<epoch()> on the column value.

=head1 SEE ALSO

=over 4

=item * L<DBIx::Class>

=item * L<DBIx::Class::TimeStamp>

=item * L<DateTime>

=back

=head1 AUTHORS

Brian Cassidy E<lt>bricas@cpan.orgE<gt>

Adam Paynter E<lt>adapay@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2006-2012 by Brian Cassidy

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut

